src/metrics/rules/active_ruleset.go

```diff
@@ -99,43 +99,34 @@ func newActiveRuleSet(
 func (as *activeRuleSet) ForwardMatch(
 	id metricid.ID,
 	fromNanos, toNanos int64,
-	opts MatchOptions,
-) (MatchResult, error) {
-	currMatchRes, err := as.forwardMatchAt(id.Bytes(), fromNanos, opts)
+	opts *MatchOptions,
+) error {
+	err := as.forwardMatchAt(id.Bytes(), fromNanos, opts)
 	if err != nil {
-		return MatchResult{}, err
+		return err
 	}
 	var (
-		forExistingID    = metadata.StagedMetadatas{currMatchRes.forExistingID}
-		forNewRollupIDs  = currMatchRes.forNewRollupIDs
 		nextIdx          = as.nextCutoverIdx(fromNanos)
 		nextCutoverNanos = as.cutoverNanosAt(nextIdx)
-		keepOriginal     = currMatchRes.keepOriginal
 	)
 
 	for nextIdx < len(as.cutoverTimesAsc) && nextCutoverNanos < toNanos {
-		nextMatchRes, err := as.forwardMatchAt(id.Bytes(), nextCutoverNanos, opts)
+		err := as.forwardMatchAt(id.Bytes(), nextCutoverNanos, opts)
 		if err != nil {
-			return MatchResult{}, err
+			return err
 		}
-		forExistingID = mergeResultsForExistingID(forExistingID, nextMatchRes.forExistingID, nextCutoverNanos)
-		forNewRollupIDs = mergeResultsForNewRollupIDs(forNewRollupIDs, nextMatchRes.forNewRollupIDs, nextCutoverNanos)
```

Comment:
Note that all this merging now happens on the provided MatchResult instead of create an additional MatchResult and merging those together.